# <b style = "color:blue"> MOVIE RECOMMENDATION SYSTEM USING MACHINE LEARNING</b>

## BUILT BY GODSPOWER UYANGA(DATA SCIENTIST/MACHINE LEARNING ENGINEER)

# <b style ="color:blue"> PROBLEM STATEMENT</b>

<b> In This Project we will Build a Content Base Recommendation System Using Machine Learning With Python, That Will Recommend A Movie To Users Base On The Content They Watched Before</b>

# <b style ="color:blue"> WORK FLOW </b>

#### <b style ="color:blue"> DATA COLLECTION :</b>Here We will Collect Textual Movie Data, That Contains Movie Name, genre, year of production etc.
#### <b style ="color:blue"> DATA PREPROCESSING :</b>After Collecting Our Data We Will Clean And Perform Neccesary Operations
#### <b style ="color:blue"> FEATURE EXTRACTION :</b>Here We Will Convert The Textual Data To Numerical, So our Machine Can Understand
#### <b style ="color:blue"> USER INPUT: </b> Base On The User Input We Will Suggest The Movies They Are Going To Watch
#### <b style ="color:blue"> COSINE SIMILARITY ALGORITH:</b>This Algorithm Will Help Suggest The Movie That Is Similar To The User Input
#### <b style ="color:blue"> BUILD RECOMMENDATION SYSTEM: </b> Here We Will Build A Recommendation System, That Will List Out Similar Movie Base On User Input.

# Importing Neccesary Dependencies

In [1]:
import numpy as np
import pandas as pd
import difflib
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# <b style = "color:blue"> DATA COLLECTION </b>

### Loading Movie Dataset

In [2]:
movies_data = pd.read_csv("movies.csv")

### Let's Peruse The First 5 Rows Of Our Movie Data

In [3]:
movies_data.head()

index     budget                                    genres  \
0      0  237000000  Action Adventure Fantasy Science Fiction   
1      1  300000000                  Adventure Fantasy Action   
2      2  245000000                    Action Adventure Crime   
3      3  250000000               Action Crime Drama Thriller   
4      4  260000000          Action Adventure Science Fiction   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  culture clash future space war space colony so...                en   
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577  ...   162.0   
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
0  Sam Worthington Zoe Saldana Sigourney Weaver S...   
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
0  [{'name': 'Stephen E. Rivkin', 'gender': 0, 'd...      James Cameron  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[5 rows x 24 columns]

### Know The Shape Of The Movie Data

In [4]:
print(f" We Have {movies_data.shape} Rows And Columns In The Data")

 We Have (4803, 24) Rows And Columns In The Data


#### We Have 4803 Rows And 24 Columns In The Movie Data And We will Extract The Important Futures We Will Use To Build Our System.

### Checking If They Is Any Null Value

In [5]:
movies_data.isnull().sum()

index                      0
budget                     0
genres                    28
homepage                3091
id                         0
keywords                 412
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title                      0
vote_average               0
vote_count                 0
cast                      43
crew                       0
director                  30
dtype: int64

### Selecting Important Features For Recommendation

In [6]:
important_features = ['genres','keywords','tagline','cast','director']
print(important_features)

['genres', 'keywords', 'tagline', 'cast', 'director']


### Replacing The Null Values With New Strings

In [7]:
for features in important_features:
    movies_data[features] = movies_data[features].fillna('')

### Adding All Our Selected Important Features Together

In [8]:
Added_features = movies_data['genres']+' '+movies_data['keywords']+' '+movies_data['tagline']+' '+movies_data['cast']+' '+movies_data['director']

### Lets Confirm Our Added Features

In [9]:
print(Added_features)

0       Action Adventure Fantasy Science Fiction cultu...
1       Adventure Fantasy Action ocean drug abuse exot...
2       Action Adventure Crime spy based on novel secr...
3       Action Crime Drama Thriller dc comics crime fi...
4       Action Adventure Science Fiction based on nove...
                              ...                        
4798    Action Crime Thriller united states\u2013mexic...
4799    Comedy Romance  A newlywed couple's honeymoon ...
4800    Comedy Drama Romance TV Movie date love at fir...
4801      A New Yorker in Shanghai Daniel Henney Eliza...
4802    Documentary obsession camcorder crush dream gi...
Length: 4803, dtype: object


### Check  Our Movie Unique Features

In [10]:
Added_features.unique()

array(['Action Adventure Fantasy Science Fiction culture clash future space war space colony society Enter the World of Pandora. Sam Worthington Zoe Saldana Sigourney Weaver Stephen Lang Michelle Rodriguez James Cameron',
       "Adventure Fantasy Action ocean drug abuse exotic island east india trading company love of one's life At the end of the world, the adventure begins. Johnny Depp Orlando Bloom Keira Knightley Stellan Skarsg\\u00e5rd Chow Yun-fat Gore Verbinski",
       'Action Adventure Crime spy based on novel secret agent sequel mi6 A Plan No One Escapes Daniel Craig Christoph Waltz L\\u00e9a Seydoux Ralph Fiennes Monica Bellucci Sam Mendes',
       ...,
       'Comedy Drama Romance TV Movie date love at first sight narration investigation team  Eric Mabius Kristin Booth Crystal Lowe Geoff Gustafson Benjamin Hollingsworth Scott Smith',
       '  A New Yorker in Shanghai Daniel Henney Eliza Coupe Bill Paxton Alan Ruck Zhu Shimao Daniel Hsia',
       'Documentary obsession camc

### Converting Our Textual Data To Feature Vector 

In [11]:
vectorizer = TfidfVectorizer()

### Confirming Our Model

In [12]:
vectorizer

TfidfVectorizer()

### Lets Transform Our Added Feature With Our Vectorizer

In [13]:
feature_vectors = vectorizer.fit_transform(Added_features)

In [14]:
print(feature_vectors)

  (0, 2432)	0.17272411194153
  (0, 7755)	0.1128035714854756
  (0, 13024)	0.1942362060108871
  (0, 10229)	0.16058685400095302
  (0, 8756)	0.22709015857011816
  (0, 14608)	0.15150672398763912
  (0, 16668)	0.19843263965100372
  (0, 14064)	0.20596090415084142
  (0, 13319)	0.2177470539412484
  (0, 17290)	0.20197912553916567
  (0, 17007)	0.23643326319898797
  (0, 13349)	0.15021264094167086
  (0, 11503)	0.27211310056983656
  (0, 11192)	0.09049319826481456
  (0, 16998)	0.1282126322850579
  (0, 15261)	0.07095833561276566
  (0, 4945)	0.24025852494110758
  (0, 14271)	0.21392179219912877
  (0, 3225)	0.24960162956997736
  (0, 16587)	0.12549432354918996
  (0, 14378)	0.33962752210959823
  (0, 5836)	0.1646750903586285
  (0, 3065)	0.22208377802661425
  (0, 3678)	0.21392179219912877
  (0, 5437)	0.1036413987316636
  :	:
  (4801, 17266)	0.2886098184932947
  (4801, 4835)	0.24713765026963996
  (4801, 403)	0.17727585190343226
  (4801, 6935)	0.2886098184932947
  (4801, 11663)	0.21557500762727902
  (4801, 1672

#### We have Finally Converted Textual Movie Data To Feature Vector

### Check The Shape Of The Feature Vectors

In [15]:
feature_vectors.shape

(4803, 17318)

#### The Feature Vectors Has  4803 Rows And 17318 Columns

## Cosine Similarity Algorithm

### Using Cosine Similarity To Obtain Similarity Scores

In [16]:
similarity_score = cosine_similarity(feature_vectors)

In [17]:
print(similarity_score)

[[1.         0.07219487 0.037733   ... 0.         0.         0.        ]
 [0.07219487 1.         0.03281499 ... 0.03575545 0.         0.        ]
 [0.037733   0.03281499 1.         ... 0.         0.05389661 0.        ]
 ...
 [0.         0.03575545 0.         ... 1.         0.         0.02651502]
 [0.         0.         0.05389661 ... 0.         1.         0.        ]
 [0.         0.         0.         ... 0.02651502 0.         1.        ]]


### Check The Shape Of Similarity Score

In [18]:
similarity_score.shape

(4803, 4803)

#### Now The Rows And Columns Of Similarity Score Is Same

### Lets Create a Lists That Contains The Names Of all The Movie Names In The Movie Dataset

In [19]:
List_Movie_Titles = movies_data['title'].tolist()
print(List_Movie_Titles)

['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre', 'The Dark Knight Rises', 'John Carter', 'Spider-Man 3', 'Tangled', 'Avengers: Age of Ultron', 'Harry Potter and the Half-Blood Prince', 'Batman v Superman: Dawn of Justice', 'Superman Returns', 'Quantum of Solace', "Pirates of the Caribbean: Dead Man's Chest", 'The Lone Ranger', 'Man of Steel', 'The Chronicles of Narnia: Prince Caspian', 'The Avengers', 'Pirates of the Caribbean: On Stranger Tides', 'Men in Black 3', 'The Hobbit: The Battle of the Five Armies', 'The Amazing Spider-Man', 'Robin Hood', 'The Hobbit: The Desolation of Smaug', 'The Golden Compass', 'King Kong', 'Titanic', 'Captain America: Civil War', 'Battleship', 'Jurassic World', 'Skyfall', 'Spider-Man 2', 'Iron Man 3', 'Alice in Wonderland', 'X-Men: The Last Stand', 'Monsters University', 'Transformers: Revenge of the Fallen', 'Transformers: Age of Extinction', 'Oz: The Great and Powerful', 'The Amazing Spider-Man 2', 'TRON: Legacy', 'Cars 2', 'Green Lant

### Getting Movie Name Input From User

In [20]:
movie_name = input("Enter Your Best Movie : ")

Enter Your Best Movie : The Craft


### Lets Find The Close Match From Our User Inputed Movie Name

In [21]:
close_match = difflib.get_close_matches(movie_name, List_Movie_Titles)
print(close_match)

['The Craft', 'The Wraith', 'The Pirate']


### Find Close Match Index

In [22]:
close_match_index = close_match[0]
print(close_match_index)

The Craft


### Finding Movie Index With Title

In [23]:
movie_Index = movies_data[movies_data.title == close_match_index]['index'].values[0]
print(movie_Index)

2568


### Getting A List Of Similar Movies From User Input

In [24]:
similarity_score_list = list(enumerate(similarity_score[movie_Index]))
print(similarity_score_list)

[(0, 0.01677933762297498), (1, 0.023132936478147514), (2, 0.0), (3, 0.015064980216668999), (4, 0.026969943091863735), (5, 0.017382080100386495), (6, 0.06889485679599461), (7, 0.0), (8, 0.10737480711241786), (9, 0.011751909761208103), (10, 0.018872891906161905), (11, 0.004944320308335385), (12, 0.045876019827675064), (13, 0.004482524619631875), (14, 0.016670001103877067), (15, 0.037821446687682274), (16, 0.0), (17, 0.011639875560499363), (18, 0.004645677486030502), (19, 0.01876772745851308), (20, 0.07484530786686791), (21, 0.023244624891627704), (22, 0.01515447644607194), (23, 0.01871380398050019), (24, 0.034781859210186054), (25, 0.008688744987997262), (26, 0.0), (27, 0.009499563999516158), (28, 0.010429400230770545), (29, 0.005823279477339969), (30, 0.012010439267941193), (31, 0.033221269166258754), (32, 0.030413260166709987), (33, 0.005582998464730167), (34, 0.0), (35, 0.0), (36, 0.0), (37, 0.1378217209019843), (38, 0.071741029809943), (39, 0.004579266880632524), (40, 0.0051519657988

### Lets Check The Size Of Our Similarity Score List

In [25]:
len(similarity_score_list)

4803

### Sorting The Movies Based On Their Similarity Score

In [26]:
sorted_similar_movies = sorted (similarity_score_list,  key = lambda x:x[1], reverse = True) 
print(sorted_similar_movies)

[(2568, 1.0000000000000002), (197, 0.17562893769451152), (1238, 0.1400186345921047), (37, 0.1378217209019843), (899, 0.133939558196388), (523, 0.1295168734402435), (1764, 0.12748052379927616), (937, 0.12530344075871766), (791, 0.12245828177704414), (743, 0.12219867744342643), (1469, 0.12164690117111303), (296, 0.12038112825895979), (331, 0.11838118542759689), (3364, 0.11694839294698885), (103, 0.11496946322127316), (109, 0.11421536955692295), (1974, 0.11391763611005752), (3062, 0.11379403005066732), (356, 0.11346338600986655), (2188, 0.11293613374658241), (390, 0.11290486793330566), (1946, 0.11177106430330108), (904, 0.11113493854191606), (113, 0.11065034593145996), (1219, 0.10987782060610807), (276, 0.10927757906468269), (8, 0.10737480711241786), (2366, 0.10599650145738065), (1695, 0.105304101546864), (565, 0.10472248756184305), (2195, 0.10463437697399013), (868, 0.10449621267186189), (464, 0.10398542327190545), (3134, 0.10380268517291125), (2763, 0.10284704550088539), (1352, 0.100665

### Printing Similar Movie Name Base On Index

In [27]:
print('HEY HERE IS OUR SUGGESTED MOVIE, YOU WILL LIKE THEM JUST CLICK ON YOUR FAVORITE: \n')

i = 1

for movie in sorted_similar_movies:
    index = movie[0]
    title_from_index = movies_data[movies_data.index==index]['title'].values[0]
    if (i<30):
        print(i, '.',title_from_index)
        i+=1

HEY HERE IS OUR SUGGESTED MOVIE, YOU WILL LIKE THEM JUST CLICK ON YOUR FAVORITE: 

1 . The Craft
2 . Harry Potter and the Philosopher's Stone
3 . The Island of Dr. Moreau
4 . Oz: The Great and Powerful
5 . Shrek
6 . Vertical Limit
7 . Return to Oz
8 . Hansel & Gretel: Witch Hunters
9 . Goosebumps
10 . Practical Magic
11 . Chill Factor
12 . End of Days
13 . Seventh Son
14 . Warlock
15 . The Sorcerer's Apprentice
16 . The Chronicles of Narnia: The Voyage of the Dawn Treader
17 . Girl, Interrupted
18 . Deuces Wild
19 . Sherlock Holmes
20 . Paparazzi
21 . Hotel Transylvania
22 . The Bad Lieutenant: Port of Call - New Orleans
23 . As Good as It Gets
24 . Harry Potter and the Order of the Phoenix
25 . Scream 4
26 . Harry Potter and the Chamber of Secrets
27 . Harry Potter and the Half-Blood Prince
28 . Dragonslayer
29 . Aladdin


## <b style="color:blue"> MOVIE RECOMMENDATION SYSTEM </b>

In [34]:
User_Name =  input("WHAT IS YOUR NAME : ")

print("-------------------------------------------------------------\n")
movie_name = input("WHAT IS THE NAME OF YOUR DESIRED MOVIE : ")

print("--------------------------------------------------------------\n")

List_Movie_Titles = movies_data['title'].tolist()

close_match = difflib.get_close_matches(movie_name, List_Movie_Titles)

close_match_index = close_match[0]

movie_Index = movies_data[movies_data.title == close_match_index]['index'].values[0]

similarity_score_list = list(enumerate(similarity_score[movie_Index]))

sorted_similar_movies = sorted (similarity_score_list,  key = lambda x:x[1], reverse = True)

print(f'{User_Name} CHECK RELATED MOVIES BELOW : \n')

i = 1

for movie in sorted_similar_movies:
    index = movie[0]
    title_from_index = movies_data[movies_data.index==index]['title'].values[0]
    if (i<30):
        print(i, '.',title_from_index)
        i+=1

WHAT IS YOUR NAME : GODSPOWER UYANGA
-------------------------------------------------------------

WHAT IS THE NAME OF YOUR DESIRED MOVIE : Eye of the Beholder
--------------------------------------------------------------

GODSPOWER UYANGA CHECK RELATED MOVIES BELOW : 

1 . Eye of the Beholder
2 . The Adventures of Pinocchio
3 . Stormbreaker
4 . Mindhunters
5 . Easy Virtue
6 . Beginners
7 . Our Kind of Traitor
8 . Die Another Day
9 . The Silence of the Lambs
10 . Miss Potter
11 . Club Dread
12 . Savages
13 . Notes on a Scandal
14 . The Village
15 . Switchback
16 . The Ghost Writer
17 . D-Tox
18 . Jane Got a Gun
19 . Lifeforce
20 . Ruby in Paradise
21 . Haywire
22 . The Cell
23 . Someone Like You...
24 . Beastly
25 . Sorority Row
26 . Where the Heart Is
27 . The Sentinel
28 . Gettysburg
29 . Angels & Demons


### Lets Check The Performance Of Our Recommendation System By Choosing Another Movie

In [35]:
User_Name =  input("WHAT IS YOUR NAME : ")

print("-------------------------------------------------------------\n")
movie_name = input("WHAT IS THE NAME OF YOUR DESIRED MOVIE : ")

print("--------------------------------------------------------------\n")

List_Movie_Titles = movies_data['title'].tolist()

close_match = difflib.get_close_matches(movie_name, List_Movie_Titles)

close_match_index = close_match[0]

movie_Index = movies_data[movies_data.title == close_match_index]['index'].values[0]

similarity_score_list = list(enumerate(similarity_score[movie_Index]))

sorted_similar_movies = sorted (similarity_score_list,  key = lambda x:x[1], reverse = True)

print(f'{User_Name} CHECK RELATED MOVIES BELOW : \n')

i = 1

for movie in sorted_similar_movies:
    index = movie[0]
    title_from_index = movies_data[movies_data.index==index]['title'].values[0]
    if (i<30):
        print(i, '.',title_from_index)
        i+=1

WHAT IS YOUR NAME : FESTUS UDOCHUKWU
-------------------------------------------------------------

WHAT IS THE NAME OF YOUR DESIRED MOVIE : The Crocodile Hunter
--------------------------------------------------------------

FESTUS UDOCHUKWU CHECK RELATED MOVIES BELOW : 

1 . Crocodile Dundee
2 . Crocodile Dundee II
3 . Crocodile Dundee in Los Angeles
4 . Meet Dave
5 . 2016: Obama's America
6 . Spider-Man 2
7 . Sexy Beast
8 . Thir13en Ghosts
9 . The Hudsucker Proxy
10 . Avatar
11 . All Is Bright
12 . Pocahontas
13 . The Borrowers
14 . My Big Fat Greek Wedding
15 . Beneath the Planet of the Apes
16 . A LEGO Brickumentary
17 . 1408
18 . A Walk Among the Tombstones
19 . A Hard Day's Night
20 . Dances with Wolves
21 . Analyze This
22 . TMNT
23 . Four Rooms
24 . My Best Friend's Wedding
25 . Where the Wild Things Are
26 . Garden State
27 . Home Alone 2: Lost in New York
28 . Nighthawks
29 . How to Lose a Guy in 10 Days


### Choose Another Movie

In [36]:
User_Name =  input("WHAT IS YOUR NAME : ")

print("-------------------------------------------------------------\n")
movie_name = input("WHAT IS THE NAME OF YOUR DESIRED MOVIE : ")

print("--------------------------------------------------------------\n")

List_Movie_Titles = movies_data['title'].tolist()

close_match = difflib.get_close_matches(movie_name, List_Movie_Titles)

close_match_index = close_match[0]

movie_Index = movies_data[movies_data.title == close_match_index]['index'].values[0]

similarity_score_list = list(enumerate(similarity_score[movie_Index]))

sorted_similar_movies = sorted (similarity_score_list,  key = lambda x:x[1], reverse = True)

print(f'{User_Name} CHECK RELATED MOVIES BELOW : \n')

i = 1

for movie in sorted_similar_movies:
    index = movie[0]
    title_from_index = movies_data[movies_data.index==index]['title'].values[0]
    if (i<30):
        print(i, '.',title_from_index)
        i+=1

WHAT IS YOUR NAME : VICTOR UYANGA
-------------------------------------------------------------

WHAT IS THE NAME OF YOUR DESIRED MOVIE : 40 Days and 40 Nights
--------------------------------------------------------------

VICTOR UYANGA CHECK RELATED MOVIES BELOW : 

1 . 40 Days and 40 Nights
2 . A Knight's Tale
3 . The Rules of Attraction
4 . The Order
5 . The Lovers
6 . Undiscovered
7 . True Lies
8 . O
9 . Here On Earth
10 . Baggage Claim
11 . Wristcutters: A Love Story
12 . Sinister 2
13 . Hollywood Homicide
14 . Resurrecting the Champ
15 . Silent House
16 . In the Valley of Elah
17 . One Missed Call
18 . The Virgin Suicides
19 . Hudson Hawk
20 . Don Jon
21 . Wicker Park
22 . 30 Days of Night
23 . Pearl Harbor
24 . True Romance
25 . Disclosure
26 . On the Line
27 . Shame
28 . You Don't Mess with the Zohan
29 . Because I Said So


# <b style ="color:blue"> SUMMARY </b>

<b style = "color:blue"> Notice That WE Have Succeeded In Bulding A Content Base Movie Recommendation System That Is Able To Recommend Movie To User Base On The User Input </b>